Add reconciler logic for DomainMapping AutoTLS

tcnghia opened the pull request:
Part of #10247 

<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Inject in a CertificateInformer
* Add reconcile logic to handle ACME challenge and Secret propagation into Ingress

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
Add AutoTLS support for DomainMapping
```

tcnghia commented:
/test pull-knative-serving-istio-stable-no-mesh-tls

tcnghia commented:
/test pull-knative-serving-istio-stable-no-mesh-tls

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +135,64 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+		return false
+	}
+	logger := logging.FromContext(ctx)
+	annotationValue := dm.Annotations[networking.DisableAutoTLSAnnotationKey]
+	disabledByAnnotation, err := strconv.ParseBool(annotationValue)
+	if annotationValue != "" && err != nil {
+		// validation should've caught an invalid value here.
+		// if we have one anyways, assume not disabled and log a warning.
+		logger.Warnf("Invalid annotation value for %q. Value: %q",
+			networking.DisableAutoTLSAnnotationKey, annotationValue)
```
this looks like a method on DM itself?

tcnghia replied:
Sure, I'll follow up.

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -37,30 +44,43 @@ import (
+// Check that our Reconciler implements CertificateAccessor
+var _ networkaccessor.CertificateAccessor = (*Reconciler)(nil)
+
+// GetNetworkingClient implements networking.CertificateAccessor
```
this is unused?

tcnghia replied:
It is used here https://github.com/knative/serving/pull/10467/files/76b30367936cb4cfdbca4446239dbde73b9fa473#diff-ebdda38d8b796c4f5e3b7e5ed54460483d22b3fd1d273de3b91a739c93aa59d6R168

vagababov replied:
I don't see a call to GetNetworkingClient there.

vagababov replied:
Hm, it's passed down... not sure we should do pass reconciler down. Rather we should pass down the networking client.

tcnghia replied:
I can follow up with a change to fix route and domainmapping reconciler both.

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -37,30 +44,43 @@ import (
+	return r.netclient
+}
+
+// GetCertificateLister implements networking.CertificateAccessor
```
ditto?

tcnghia replied:
It is used here https://github.com/knative/serving/pull/10467/files/76b30367936cb4cfdbca4446239dbde73b9fa473#diff-ebdda38d8b796c4f5e3b7e5ed54460483d22b3fd1d273de3b91a739c93aa59d6R168

tcnghia commented:
/assign @vagababov @YoussB

tcnghia commented:
/assign @julz

julz commented on pkg/apis/serving/v1alpha1/domainmapping_lifecycle.go:
```diff
@@ -62,7 +62,7 @@ const (
 	// DomainMappingConditionCertificateProvisioned condition when it is set to True
 	// because AutoTLS was not enabled.
-	AutoTLSNotEnabledMessage = "autoTLS is not enabled"
+	AutoTLSNotEnabledMessage = "AutoTLS is not enabled"
```
should we update the corresponding message in [route_lifecyce.go](https://github.com/knative/serving/blob/3695f5b03a9cc4dc2ef83741e0481f4c95e98094/pkg/apis/serving/v1/route_lifecycle.go#L174) to match?

tcnghia replied:
I will follow up by making both changes in the same PR. for now I reverted this line.

tcnghia commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -19,12 +19,19 @@ package domainmapping
+	kaccessor "knative.dev/serving/pkg/reconciler/accessor"
+	networkaccessor "knative.dev/serving/pkg/reconciler/accessor/networking"
+
+	"github.com/prometheus/common/log"
```
Nice catch. I removed.

tcnghia commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +135,64 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+	acmeChallenges := []netv1alpha1.HTTP01Challenge{}
+	desiredCert := resources.MakeCertificate(dm, certClass(ctx))
+	cert, err := networkaccessor.ReconcileCertificate(ctx, dm, desiredCert, r)
+	log.Infof("Found the cert: %+v", cert)
```
Nice catch. I removed.

tcnghia commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +135,64 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+	tls := []netv1alpha1.IngressTLS{}
+	if !autoTLSEnabled(ctx, dm) {
+		dm.Status.MarkTLSNotEnabled(v1.AutoTLSNotEnabledMessage)
+		return tls, nil, nil
```
Changed to `nil, nil, nil` for consistency

tcnghia commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +135,64 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+		acmeChallenges = append(acmeChallenges, cert.Status.HTTP01Challenges...)
+		dm.Status.MarkCertificateNotReady(cert.Name)
+	}
+	sort.Slice(acmeChallenges, func(i, j int) bool {
```
The challenges were populated by the certificate controller after we create the cert. I think we can require that on the Cert controller side, but may still be good to be defensive here.

tcnghia commented:
/test pull-knative-serving-istio-stable-no-mesh

tcnghia commented:
/hold

tcnghia commented:
/unhold
(tested using latest version of net-http01)

@julz @vagababov I've addressed most of your feedbacks, except for a few that I will follow up.

julz commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +134,63 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+}
+
+func (r *Reconciler) tls(ctx context.Context, dm *v1alpha1.DomainMapping) ([]netv1alpha1.IngressTLS, []netv1alpha1.HTTP01Challenge, error) {
+	tls := []netv1alpha1.IngressTLS{}
```
nit: I'd move this lower down (or just do it inline on L185) since I think there's only one append here now

tcnghia replied:
Done

julz commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +134,63 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+	}
+	if cert.IsReady() {
+		dm.Status.MarkCertificateReady(cert.Name)
+		tls = append(tls, routeresources.MakeIngressTLS(cert, dnsNames.List()))
```
I think you could directly do 

```suggestion
		return []netv1alpha1.IngressTLS{routeresources.MakeIngressTLS(cert, dnsNames.List()))}, nil, nil
```

here, and then the else can be dropped and out-dented.

tcnghia replied:
Done

julz reviewed (approved):
/lgtm
/hold so @vagababov can have a look and in case you want to do the little suggested refactor (but Im fine if you prefer this way or want to do it separately).

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +134,62 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+	if annotationValue != "" && err != nil {
+		// Validation should've caught an invalid value here.
+		// If we have one anyway, assume not disabled and log a warning.
+		logger.Warnf("Invalid annotation value for %q. Value: %q",
```
```suggestion
		logger.Warnf("DM.Annotations[%s] = %q is invalid",
```

tcnghia replied:
Done

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +134,62 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+	if !config.FromContext(ctx).Network.AutoTLS {
+		return false
+	}
+	logger := logging.FromContext(ctx)
```
Move logger in the `if` since it's unused otherwise.

tcnghia replied:
Done

vagababov commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -110,6 +134,62 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, dm *v1alpha1.DomainMappi
+		return nil, nil, err
+	}
+
+	dnsNames := sets.NewString(desiredCert.Spec.DNSNames...)
+	if dnsNames.Has(dm.Name) {
+		dm.Status.URL.Scheme = "https"
+	}
```
So you create a map and iterate over all the elements just to see if the item is in the list. Well we can check that instead of iterating over the map.

And below, we can probably just inject `desiredCert.Spec.DNSNames`.

tcnghia replied:
Done

vagababov commented on pkg/reconciler/domainmapping/reconciler_test.go:
```diff
@@ -0,0 +1,80 @@
+				},
+			})
+
+			dm.Annotations[networking.DisableAutoTLSAnnotationKey] = tc.tlsDisabledAnnotation
```
this does not cover `dm.Annotations==nil` case.

tcnghia replied:
Done

vagababov commented on pkg/reconciler/domainmapping/reconciler_test.go:
```diff
@@ -62,16 +61,25 @@ func TestAutoTLSEnabled(t *testing.T) {
-			dm.Annotations[networking.DisableAutoTLSAnnotationKey] = tc.tlsDisabledAnnotation
-
+			if tc.tlsDisabledAnnotation != "" {
+				dm.Annotations = map[string]string{}
+				dm.Annotations[networking.DisableAutoTLSAnnotationKey] = tc.tlsDisabledAnnotation
+			}
```
```suggestion
				dm.Annotations = map[string]string{
				 networking.DisableAutoTLSAnnotationKey: tc.tlsDisabledAnnotation,
				 }
			}
```

vagababov commented:
/lgtm
/unhold

mattmoor commented on pkg/reconciler/domainmapping/reconciler.go:
```diff
@@ -37,30 +42,43 @@ import (
 	domainmappingreconciler "knative.dev/serving/pkg/client/injection/reconciler/serving/v1alpha1/domainmapping"
 	"knative.dev/serving/pkg/reconciler/domainmapping/config"
 	"knative.dev/serving/pkg/reconciler/domainmapping/resources"
+	routeresources "knative.dev/serving/pkg/reconciler/route/resources"
```
Can we move this somewhere common?

tcnghia replied:
I'll follow up with that as well.

n3wscott replied:
Got an issue with a DomainMapping:
```
party       knative.party                                Unknown   NewObservedGenFailure
```

Looking at the logs of the kcert:

```
  Warning  InternalError  3m23s (x17 over 8m51s)  certificate-controller  Service "knative.party" is invalid: metadata.name: Invalid value: "knative.party": a DNS-1035 label must consist of lower case alphanumeric characters or '-', start with an alphabetic character, and end with an alphanumeric character (e.g. 'my-name',  or 'abc-123', regex used for validation is '[a-z]([-a-z0-9]*[a-z0-9])?')```

mattmoor replied:
I believe you need to update net-http01, let's discuss in slack

mattmoor reviewed (commented):
/lgtm
